Add basic implementation of steno order to editing table.

sofa13 opened the pull request:
### About

Implemented basic steno order by enumerating every letter in the stroke to a letter in the english language alphabet order, so that the builtin sorting function for the table does the rest of the sorting! A couple of exceptions were handled, I'm sure there are more to point out.
### Issues

Fixes #1
### Tested Platforms

Tested on Windows 7.

Commit:
Add basic implementation of steno order to editing table.

morinted commented:
Your alphabet system is clever, but there are some concerns I have:

- We need to be generic -- in the case of the Palantype there are over 30 keys and therefore this system won't work for larger keyboards. If I remember correctly Kinglet uses a full keyboard, something like 50-60 keys.
- Numbers are a reality in Plover. Some strokes are normalized to numbers, e.g. `#STPHAOEU` might come as `123450EU`. We should inspect the stroke object and see what we have to work with.
- I'd throw this sort function into the Stroke class so that it can be used elsewhere in the application.

We can brainstorm on this in person.

morinted commented:
While debugging, it looks like the steno is actually just:

<img width="419" alt="screen shot 2016-09-26 at 10 35 52 am" src="https://cloud.githubusercontent.com/assets/5840970/18838358/30768b46-83d5-11e6-8bb3-1e745f0664bd.png">

I'd like to look and see if we can get access to a Stroke object.

morinted commented:
Over in json_dict, you can see that while loading a steno dictionary we do normalize the strokes:

<img width="588" alt="screen shot 2016-09-26 at 10 39 14 am" src="https://cloud.githubusercontent.com/assets/5840970/18838451/8a9695d0-83d5-11e6-925b-c0c06c5e8385.png">

morinted commented:
So, given that the steno is normalized, there are two cases I can see:

- Use the fact that the steno is normalized to your advantage: divide the stroke by its hyphen or by the implicit hyphen keys, then first check if the left keys have different indexes in the system key order, then do the same for the center keys, then the right keys.
- We change to storing Stroke and Translation objects -- then use the explicit keys (formatted as 'S-', 'T-', etc.) to sort.

morinted commented:
[This Stackoverflow answer](http://stackoverflow.com/a/24782540) has some tips for lexigraphical sorting.

This is what you get from the system exports:

```
NUMBERS {'-P': '-7', '-T': '-9', 'O-': '0-', '-L': '-8', 'T-': '2-', '-F': '-6', 'S-': '1-', 'P-': '3-', 'H-': '4-', 'A-': '5-'}
NUMBER_KEY #
ORTHOGRAPHY_RULES [(re.compile('^(.*[aeiou]c) \\^ ly$', re.IGNORECASE), '\\1ally'), (re.compile('^(.*t)e \\^ ry$', re.IGNORECASE), '\\1ory'), (re.compile('^(.*[naeiou])te? \\^ cy$', re.IGNORECASE), '\\1cy'), (re.compile('^(.*(?:s|sh|x|z|zh)) \\^ s$', re.IGNORECASE), '\\1es'), (re.compile('^(.*(?:oa|ea|i|ee|oo|au|ou|l|n|(?<![gin]a)r|t)ch) \\^ s$', re.IGNORECASE), '\\1es'), (re.compile('^(.+[bcdfghjklmnpqrstvwxz])y \\^ s$', re.IGNORECASE), '\\1ies'), (re.compile('^(.+)ie \\^ ing$', re.IGNORECASE), '\\1ying'), (re.compile('^(.+[cdfghlmnpr])y \\^ ist$', re.IGNORECASE), '\\1ist'), (re.compile('^(.+[bcdfghjklmnpqrstvwxz])y \\^ ([a-hj-xz].*)$', re.IGNORECASE), '\\1i\\2'), (re.compile('^(.+)te \\^ en$', re.IGNORECASE), '\\1tten'), (re.compile('^(.+e)e \\^ (e.+)$', re.IGNORECASE), '\\1\\2'), (re.compile('^(.+[bcdfghjklmnpqrstuvwxz])e \\^ ([aeiouy].*)$', re.IGNORECASE), '\\1\\2'), (re.compile('^(.*(?:[bcdfghjklmnprstvwxyz]|qu)[aeiou])([bcdfgklmnprtvz]) \\^ ([aeiouy].*)$', re.IGNORECASE), '\\1\\2\\2\\3')]
UNDO_STROKE_STENO *
KEYMAPS {'ProCAT': {'W-': 'W-', '-P': '-P', '-D': '-D', 'O-': 'O-', '-L': '-L', '-U': '-U', '-F': '-F', '-G': '-G', '#': '#', '-S': '-S', '-E': '-E', '-T': '-T', 'K-': 'K-', '*': '*', 'T-': 'T-', '-Z': '-Z', 'A-': 'A-', '-R': '-R', 'S-': 'S-', 'R-': 'R-', 'P-': 'P-', 'H-': 'H-', '-B': '-B'}, 'Gemini PR': {'W-': 'W-', '-P': '-P', '-D': '-D', 'O-': 'O-', '-L': '-L', '-U': '-U', '-F': '-F', 'no-op': ('Fn', 'pwr', 'res1', 'res2'), '-G': '-G', '#': ('#1', '#2', '#3', '#4', '#5', '#6', '#7', '#8', '#9', '#A', '#B', '#C'), '-S': '-S', '-E': '-E', '-T': '-T', 'K-': 'K-', '*': ('*1', '*2', '*3', '*4'), 'T-': 'T-', '-Z': '-Z', 'A-': 'A-', '-R': '-R', 'S-': ('S1-', 'S2-'), 'R-': 'R-', 'P-': 'P-', 'H-': 'H-', '-B': '-B'}, 'Keyboard': {'W-': 'd', '-P': 'i', '-D': '[', 'O-': 'v', '-L': 'o', '-U': 'm', '-F': 'u', 'no-op': ('z', 'x', 'b', ',', '.', '/', ']', '\\'), '-G': 'l', '#': ('1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '='), '-S': ';', '-E': 'n', '-T': 'p', 'K-': 's', 'arpeggiate': 'space', '*': ('t', 'g', 'y', 'h'), 'T-': 'w', '-Z': "'", 'A-': 'c', '-R': 'j', 'S-': ('a', 'q'), 'R-': 'f', 'P-': 'e', 'H-': 'r', '-B': 'k'}, 'TX Bolt': {'W-': 'W-', '-P': '-P', '-D': '-D', 'O-': 'O-', '-L': '-L', '-U': '-U', '-F': '-F', '-G': '-G', '#': '#', '-S': '-S', '-E': '-E', '-T': '-T', 'K-': 'K-', '*': '*', 'T-': 'T-', '-Z': '-Z', 'A-': 'A-', '-R': '-R', 'S-': 'S-', 'R-': 'R-', 'P-': 'P-', 'H-': 'H-', '-B': '-B'}, 'Stentura': {'W-': 'W-', '-P': '-P', '-D': '-D', 'O-': 'O-', '-L': '-L', '-U': '-U', '-F': '-F', 'no-op': '^', '-G': '-G', '#': '#', '-S': '-S', '-E': '-E', '-T': '-T', 'K-': 'K-', '*': '*', 'T-': 'T-', '-Z': '-Z', 'A-': 'A-', '-R': '-R', 'S-': 'S-', 'R-': 'R-', 'P-': 'P-', 'H-': 'H-', '-B': '-B'}, 'Passport': {'W-': 'W', '-P': 'N', '-D': 'D', 'O-': 'O', '-L': 'L', '-U': 'U', '-F': 'F', 'no-op': ('!', '^', '+'), '-G': 'G', '#': '#', '-S': 'X', '-E': 'E', '-T': 'Y', 'K-': 'K', '*': ('~', '*'), 'T-': 'T', '-Z': 'Z', 'A-': 'A', '-R': 'Q', 'S-': ('S', 'C'), 'R-': 'R', 'P-': 'P', 'H-': 'H', '-B': 'B'}, 'Treal': {'W-': 'W-', '-P': '-P', '-D': '-D', 'O-': 'O-', '-L': '-L', '-U': '-U', '-F': '-F', 'no-op': ('X1-', 'X2-', 'X3'), '-G': '-G', '#': ('#1', '#2', '#3', '#4', '#5', '#6', '#7', '#8', '#9', '#A', '#B'), '-S': '-S', '-E': '-E', '-T': '-T', 'K-': 'K-', '*': ('*1', '*2'), 'T-': 'T-', '-Z': '-Z', 'A-': 'A-', '-R': '-R', 'S-': ('S1-', 'S2-'), 'R-': 'R-', 'P-': 'P-', 'H-': 'H-', '-B': '-B'}}
SUFFIX_KEYS {'-G', '-Z', '-D', '-S'}
ORTHOGRAPHY_RULES_ALIASES {'able': 'ible'}
KEYS ('#', 'S-', 'T-', 'K-', 'P-', 'W-', 'H-', 'R-', 'A-', 'O-', '*', '-E', '-U', '-F', '-R', '-P', '-B', '-L', '-G', '-T', '-S', '-D', '-Z')
KEY_ORDER {'W-': 5, '-P': 15, '-D': 21, 'O-': 9, '-L': 17, '#': 0, '-U': 12, '-F': 13, '4-': 6, '2-': 2, '3-': 4, 'R-': 7, '-S': 20, '-E': 11, '0-': 9, '-8': 17, '-6': 13, 'K-': 3, 'H-': 6, '-G': 18, '-T': 19, '-7': 15, '*': 10, 'T-': 2, '-Z': 22, '-B': 16, '-R': 14, 'S-': 1, '5-': 8, 'P-': 4, '1-': 1, 'A-': 8, '-9': 19}
IMPLICIT_HYPHENS {'5', 'A', 'O', '0', '*', 'U', 'E'}
```

The KEY_ORDER has the indexes for order that you need. The `-` isn't a key, but will help you distinguish between R- and -R (7 and 14) or T- and -T (2 and 19). It's nice, it even has numbers.

morinted commented:
```
left side: 'x-'
So you encounter S, so you lookup 'S-'

implicit hyphen OR actual hyphen
if implicit hyphen, lookup the letter ('A', 'A-', '-A') and set side to right
if hyphen, set side to right

right side: '-x'
So you encounter another S, so you lookup '-S'
```

morinted commented:
Looks like you can hand in a cmp function to do a custom comparison :)

https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types

On Sep 25, 2016 9:47 PM, "Sophie Le Page" <notifications@github.com> wrote:

> About
>
> Implemented basic steno order by enumerating every letter in the stroke to
> a letter in the english language alphabet order, so that the builtin
> sorting function for the table does the rest of the sorting! A couple of
> exceptions were handled, I'm sure there are more to point out.
> Issues
>
> Fixes #1 <https://github.com/Germanika/plover/issues/1>
> Tested Platforms
>
> Tested on Windows 7.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Germanika/plover/pull/2
> Commit Summary
>
>    - Add basic implementation of steno order to editing table.
>
> File Changes
>
>    - *M* plover/gui_qt/dictionary_editor.py
>    <https://github.com/Germanika/plover/pull/2/files#diff-0> (46)
>
> Patch Links:
>
>    - https://github.com/Germanika/plover/pull/2.patch
>    - https://github.com/Germanika/plover/pull/2.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Germanika/plover/pull/2>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFkgSniGyQMs4Qmdm_RlJkH1coKlXmHpks5qtyQigaJpZM4KGEeX>
> .
>

panathea commented on plover/gui_qt/dictionary_editor.py:
```diff
@@ -19,10 +19,11 @@
 
 from plover.gui_qt.dictionary_editor_ui import Ui_DictionaryEditor
 from plover.gui_qt.utils import ToolBar, WindowState
+from functools import cmp_to_key
```
https://www.python.org/dev/peps/pep-0008/#imports

In Python, system imports go on top of local imports according to PEP8. To quote:

> 1. standard library imports
> 2. related third party imports
> 3. local application/library specific imports

sofa13 commented:
Updated the steno order function to use compare (scrapped the clever method D:) and implemented the compare_normalize_stroke function in plover\steno.py.

The function works pretty well (I'll need Ted to verify the subtleties), but the sorting takes wayyy too long for very large dictionaries such as main (takes about 5 seconds to sort!!!). I am working on making it faster, shouldn't be a problem! Until then I thought I'd share the simple code I've come up with as a starting point.

Thanks again to Ted for the great comments above, very helpful!

panathea commented on plover/steno.py:
```diff
@@ -21,6 +21,78 @@
+    for i in range(len(x.strokes)):
+        if i > len(y.strokes) - 1:
+            return 1
+        x_norm = normalize_stroke(x.strokes[i])
```
Since the function is "compare_normalize" I think you don't need to normalize the stroke before compare, it's implied.

panathea commented on plover/steno.py:
```diff
@@ -21,6 +21,78 @@
+    return 0
+
+def less_than(x, y):
+    x_right_side = True
```
I think this is flipped -- it should be "left side" because the left keys are:

S- T- K-... and the right keys are -F -R -P...

panathea commented on plover/steno.py:
```diff
@@ -21,6 +21,78 @@
 _IMPLICIT_NUMBER_RX = re.compile('(^|[1-4])([6-9])')
 
+def compare_normalize_stroke(x, y):
+    for i in range(len(x.strokes)):
```
My intuition here would be:

``` python
for x, y in zip(x.strokes, y.strokes):
    if less_than(x, y):
        return -1
    elif less_than(y, x):
        return 1
return cmp(len(x.strokes), len(y.strokes))
```

panathea commented on plover/steno.py:
```diff
@@ -21,6 +21,78 @@
+def less_than(x, y):
+    x_right_side = True
+    y_right_side = True
+    for i in range(len(x)):
```
`for key in x:` would suffice

panathea replied:
Oh there's danger in what I said here. I'll see if I can simplify the problem a little.

panathea replied:
Okay -- more logical now, since you have multiple levels of comparison here it may as well be represented.

I'd say name your two functions "cmp_normalized_steno" and "cmp_normalized_strokes"

This is untested (might not even run O.O)

``` python
def cmp_normalized_steno(x, y):
    for stroke_x, stroke_y in zip(x, y):
        cmp_strokes = cmp_normalized_strokes(x, y)
        if cmp_strokes:
            return cmp_strokes
    return cmp(len(x.strokes), len(y.strokes))

def cmp_normalized_strokes(x, y):
    # Turn string into a 'stack'
    x = list(reverse(x))
    y = list(reverse(y))

    # Returns key index and whether it is on the left of the board.
    def get_key_index(key, side):
        if side == 1:  # Right side, easiest case.
            return system.KEY_ORDER.get('-%s' % key), 1
        elif key == '-'  # Center delimiter
            return None, 1
        elif key in system.IMPLICIT_HYPHENS:  # Detect center keys
            index = (
                system.KEY_ORDER.get(key) or
                system.KEY_ORDER.get('%s-' % key) or
                system.KEY_ORDER.get('-%s' % key)
            )
            assert index is not None, 'Invalid implicit key found: %s' % key
            return index, 0
        elif side == 0:
            # Case where we last found a center key, but now can't.
            # It implies our first right side key.
            return system.KEY_ORDER.get('-%s' % key), 1
        else: # Must be a left-side key.
            return system.KEY_ORDER.get('%s-' % key), -1

    # -1: left, 0: center, 1: right
    x_side = -1
    y_side = -1

    # While x and y both have content...
    while x and y:
        x_index, x_side = get_key_index(x.pop(), x_side)
        if x_index is None:
            x_index, x_side = get_key_index(x.pop(), x_side)
        y_index, y_side = get_key_index(x.pop(), y_side)
        if y_index is None:
            y_index, y_side = get_key_index(x.pop(), y_side)
        index_compare = cmp(x_index, y_index)
        if index_compare: # If not equal, we have a result.
            return index_compare
    # One will be zero at least.
    return cmp(len(x), len(y))
```

panathea reviewed (commented):
Some hopefully helpful feedback for about half of what's here!

morinted commented:
I didn't realize that in Python 3 they discourage use of `cmp`. I'll try a solution with only key.

See: http://python3porting.com/preparing.html#keycmp-section

Also my tests are promising:

```python
>>> a = (0, 1, 2, 3, -1, 4)
>>> b = (0, 1, 2, 3, 0)
>>> c = (1, 2, 3)
>>> sorted([a, b, c])
[(0, 1, 2, 3, -1, 4), (0, 1, 2, 3, 0), (1, 2, 3)]
>>> d = (-1, 5)
>>> sorted([a, b, d, c])
[(-1, 5), (0, 1, 2, 3, -1, 4), (0, 1, 2, 3, 0), (1, 2, 3)]
>>> a = ((0, 1, 2, 3,), (4,))
>>> b = (b)
>>> c = (c)
>>> d = (d)
>>> d
(-1, 5)
>>> d = (d,)
>>> d
((-1, 5),)
>>> b = (b,)
>>> c = (c,)
>>> sorted([a, b, d, c])
[((-1, 5),), ((0, 1, 2, 3), (4,)), ((0, 1, 2, 3, 0),), ((1, 2, 3),)]
>>>
```

morinted commented:
Okay, still basing off of your original solution, I've implemented it as a key.

The initial load is quite slow, upwards of 5 seconds for all my dictionaries, but I store the results in a dictionary for future sorts. So, once loaded, it's not so bad afterwards.

Here's the full change that I came up with:

```python
diff --git a/plover/gui_qt/dictionary_editor.py b/plover/gui_qt/dictionary_editor.py
index 91ed54d..2c84479 100644
--- a/plover/gui_qt/dictionary_editor.py
+++ b/plover/gui_qt/dictionary_editor.py
@@ -19,7 +19,11 @@ from PyQt5.QtWidgets import (
 
 from plover.translation import escape_translation, unescape_translation
 from plover.misc import expand_path, shorten_path
-from plover.steno import normalize_steno, filter_entry
+from plover.steno import (
+    normalize_steno,
+    filter_entry,
+    normalized_steno_to_indexes
+)
 
 from plover.gui_qt.dictionary_editor_ui import Ui_DictionaryEditor
 from plover.gui_qt.utils import ToolBar, WindowState
@@ -198,7 +202,9 @@ class DictionaryItemModel(QAbstractTableModel):
 
     def sort(self, column, order):
         self.layoutAboutToBeChanged.emit()
-        if column == 2:
+        if column == 0:
+            key = lambda entry: normalized_steno_to_indexes(entry.strokes)
+        elif column == 2:
             key = attrgetter('dictionary_path')
         elif column == 3:
             key = lambda entry: len(entry.strokes)
diff --git a/plover/steno.py b/plover/steno.py
index 0570da9..41a2471 100644
--- a/plover/steno.py
+++ b/plover/steno.py
@@ -70,6 +70,52 @@ def filter_entry(strokes, translation, strokes_filter=None,
     return True
 
 
+normalized_steno_indexes = {}
+
+
+def normalized_steno_to_indexes(steno):
+    return (
+        normalized_steno_indexes[steno]
+        if steno in normalized_steno_indexes
+        else normalized_steno_indexes.setdefault(
+            steno,
+            tuple(normalized_stroke_to_indexes(stroke) for stroke in steno)
+        )
+    )
+
+
+def normalized_stroke_to_indexes(stroke):
+    side = -1
+
+    indexes = []
+    for key in stroke:
+        if side == 1:  # Right side, easiest case.
+            key_index = system.KEY_ORDER.get('-%s' % key)
+        elif key == '-':  # Center delimiter
+            side = 1
+            continue
+        elif key == system.NUMBER_KEY:
+            key_index = system.KEY_ORDER.get(key)
+        elif key in system.IMPLICIT_HYPHENS:  # Detect center keys
+            key_index = (
+                system.KEY_ORDER.get(key) or
+                system.KEY_ORDER.get('%s-' % key) or
+                system.KEY_ORDER.get('-%s' % key)
+            )
+            assert key_index is not None, 'Invalid implicit key found: %s' % key
+            side = 0
+        elif side == 0:
+            # Case where we last found a center key, but now can't.
+            # It implies our first right side key.
+            key_index = system.KEY_ORDER.get('-%s' % key)
+            side = 1
+        else: # Must be a left-side key.
+            key_index = system.KEY_ORDER.get('%s-' % key)
+        if key_index is not None:
+            indexes.append(key_index)
+    return tuple(indexes)
+
+
 class Stroke(object):
     """A standardized data model for stenotype machine strokes.
 ```

It would still be nice to add tests.

For `test_steno.py`, I'd start off with:

```python
    def test_stroke_to_indexes(self):
        cases = (
            ('STPH', (1, 2, 4, 6,)),
            ('#', (0,)),
        )
        for stroke, indexes in cases:
            self.assertEqual(indexes, normalized_stroke_to_indexes(stroke))
```

morinted commented:
Also cache the stroke function:

```python
normalized_stroke_indexes = {}


def normalized_stroke_to_indexes(stroke):
    if stroke not in normalized_stroke_indexes:
        side = -1
        indexes = []
        for key in stroke:
            if side == 1:  # Right side, easiest case.
                key_index = system.KEY_ORDER.get('-%s' % key)
            elif key == '-':  # Center delimiter
                side = 1
                continue
            elif key == system.NUMBER_KEY:
                key_index = system.KEY_ORDER.get(key)
            elif key in system.IMPLICIT_HYPHENS:  # Detect center keys
                key_index = (
                    system.KEY_ORDER.get(key) or
                    system.KEY_ORDER.get('%s-' % key) or
                    system.KEY_ORDER.get('-%s' % key)
                )
                assert key_index is not None,\
                    'Invalid implicit key found: %s' % key
                side = 0
            elif side == 0:
                # Case where we last found a center key, but now can't.
                # It implies our first right side key.
                key_index = system.KEY_ORDER.get('-%s' % key)
                side = 1
            else: # Must be a left-side key.
                key_index = system.KEY_ORDER.get('%s-' % key)
            if key_index is not None:
                indexes.append(key_index)
    return (
        normalized_stroke_indexes[stroke]
        if stroke in normalized_stroke_indexes
        else normalized_stroke_indexes.setdefault(
            stroke,
            tuple(indexes)
        )
    )
```

sofa13 commented:
Alright, thanks to Teds help the performance was greatly improved, now ~3 seconds on first load for very large dictionaries. Test cases have also been added.

panathea commented on test/test_steno.py:
```diff
@@ -59,3 +59,19 @@ def test_steno(self):
+            ('#-Z', (0, 22,)),
+            ('12K3W*', (1, 2, 3, 4, 5, 10,)),
+            ('ED', (11, 21,)),
+            ('*ED', (10, 11, 21,))
```
We should hit the edge cases, all the edges:

`#, *, S-, R-, A-, -U, -F, -Z`.

Mainly concerned with combos of most of these (using judgment)

I didn't originally write it this way, either, but we should probably put the steno through normalization first, just in case, as that is the expected behavior of the program. That way if the normalization function gets updated, the compare tests will start to fail too!

So something like:

`self.assertEqual(indexes, normalized_stroke_to_indexes(normalize_stroke(stroke)))`

sofa13 commented:
The test case has been updated, with a way to test the normalize_stroke function as well as many more test cases to hit all the edge cases as well as a few discretionary combos of these thrown in!
